physics.py

```diff
@@ -171,12 +178,96 @@ def update(self, tm_diff: float, robotVoltage: float) -> None:
+        deltaTrans = Transform2d(deltaX, -deltaY, deltaHeading)
 
         newPose = self.pose + deltaTrans
         self.pose = newPose
 
 
+class NoteSim:
+    def __init__(self) -> None:
+        self.midlineNotes = constants.kNotesStartingMidline
+        self.blueNotes = constants.kNotesStartingBlueWing
+        self.redNotes = constants.kNotesStartingRedWing
+
+        self.loadingNotes = [
+            constants.kNoteLoadingStationPositionBlue,
+            constants.kNoteLoadingStationPositionRed,
+        ]
+
+    def canPickup(self, note: Pose3d, botPose) -> bool:
+        if pointInCircle(botPose.translation(), note.toPose2d().translation(), 0.5):
+            return True
+        return False
+
+    def update(self, _tm_diff, bot: MentorBot):
+        SmartDashboard.putNumberArray(
+            constants.kSimNotePositionsKey,
+            convertToSendablePoses(
+                [
+                    *self.midlineNotes,
+                    *self.blueNotes,
+                    *self.redNotes,
+                    *self.loadingNotes,
+                ]
+            ),
+        )
+
+        # check whether intaking, update sensors according to position on field
+
+        intaking = bot.container.intake.state == IntakeSubsystem.IntakeState.Intaking
+
+        botPose = Pose2d(
+            *SmartDashboard.getNumberArray(constants.kSimRobotPoseArrayKey, [0, 0, 0])
+        )
+
+        hasNote = SmartDashboard.getBoolean(
+            f"{bot.container.intake.intakeMotor.getNettableIden()}/fwdLimit",
+            False,
+        )
+
+        if intaking:
+            notestate = hasNote
+            for stationObject in self.loadingNotes:
+                if self.canPickup(stationObject, botPose):
+                    notestate = True
+
+            for blueWingNote in self.blueNotes:
+                # remove the note from the field
+                if self.canPickup(blueWingNote, botPose):
+                    notestate = True
+                    self.blueNotes.remove(blueWingNote)
+
+            for redWingNote in self.redNotes:
+                # remove the note from the field
+                if self.canPickup(redWingNote, botPose):
+                    notestate = True
+                    self.redNotes.remove(redWingNote)
+
+            for midlineNote in self.midlineNotes:
+                # remove the note from the field
+                if self.canPickup(midlineNote, botPose):
+                    notestate = True
+                    self.midlineNotes.remove(midlineNote)
+
+            SmartDashboard.putBoolean(
+                f"{bot.container.intake.intakeMotor.getNettableIden()}/fwdLimit",
+                notestate,
+            )
+
+        # shooting a note clears the note
+        feeding = bot.container.intake.state == IntakeSubsystem.IntakeState.Feeding
+
+        if feeding:
+            if hasNote:
+                pass  # Logic for calculating a shot
```

Comment:
How did any of this end up as the advantagescope visualization?

Reply:
wdym?

Reply:
like bc in Ivan's video you can see notes flying around and stuff, how does this accomplish it?

Reply:
thats done in #41